src/mono/netcore/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Mono.cs

```diff
@@ -354,6 +356,86 @@ public static unsafe IntPtr StringToBSTR(string? s)
                 return BufferToBSTR(fixed_s, s.Length);
         }
 
+        internal class MarshalerInstanceKeyComparer : IEqualityComparer<(Type, string)>
+        {
+            public bool Equals((Type, string) lhs, (Type, string) rhs)
+            {
+                return lhs.CompareTo(rhs) == 0;
+            }
+
+            public int GetHashCode((Type, string) key)
+            {
+                return key.GetHashCode();
+            }
+        }
+
+        internal static Dictionary<(Type, string), ICustomMarshaler> MarshalerInstanceCache;
+        internal static readonly object MarshalerInstanceCacheLock = new object();
+
+        internal static ICustomMarshaler GetCustomMarshalerInstance(Type type, string cookie)
+        {
+            var key = (type, cookie);
+
+            LazyInitializer.EnsureInitialized(
+                ref MarshalerInstanceCache,
+                () => new Dictionary<(Type, string), ICustomMarshaler>(new MarshalerInstanceKeyComparer())
+            );
+
+            ICustomMarshaler result;
+            bool gotExistingInstance;
+            lock (MarshalerInstanceCacheLock)
+                gotExistingInstance = MarshalerInstanceCache.TryGetValue(key, out result);
+
+            if (!gotExistingInstance)
```

Comment:
isn't the whole block should be under `lock` ? otherwise it might be executed twice.
Also I guess you can lock on `MarshalerInstanceCache` without a separate lock object (which introduces a static constructor) since `LazyInitializer.EnsureInitialized` is thread-safe.

Reply:
Yup, the code should have extended the lock across the uses of MarshalerInstanceCache, not just when getting/setting. I will do that

Reply:
AFAICT this would only potentially cause churn from misuse of the cache and not actual bad state.

Reply:
@EgorBo hm now I am not sure if I want to "fix" this: it seems valid to avoid a potentially long wait, with the cost of some cache misses/un-needed cache refreshes/heap churn. 

I will remove the separate lock object though.